### PR TITLE
Support running commands against the previous yielded subject

### PR DESCRIPTION
**Pull request by tlrobinson**
**What**:

Changes the custom Cypress commands to respect the previously yielded subject, for example, with the following document:

```html
    <button>Button Text 1</button>
    <div id="nested">
      <button>Button Text 2</button>
    </div>
```

this

```js
cy.get('#nested').findByText('Button Text 1').should('not.exist')
```

would previously fail because `findByText` was not passed the result of `cy.get('#nested')` as its `container` (defaulting to `document`)

**Why**:

It's a convenient way to scope commands to the result of a previous command, and for users of builtin Cypress commands it's surprising that `@testing-library/cypress` doesn't currently behave like this.

**How**:

The [`prevSubject` option](https://docs.cypress.io/api/cypress-api/custom-commands.html#Usage) is now passed to `Cypress.Commands.add()` (with a value of `['optional', 'document', 'element', 'window']` which matches [builtin commands like `contains`](https://github.com/cypress-io/cypress/blob/develop/packages/driver/src/cy/commands/querying.js#L384). This causes the subject (if any) to be passed as the first argument to the command. The subject is given lower precedence than the `testing-library`'s `container` option, but higher precedence than the document itself. 

**Checklist**:

- [x] Documentation N/A
- [x] Tests
- [x] Ready to be merged

### CODE REVIEW

File: cypress/integration/find.spec.js [hunk 1]
```diff
@@ -79,27 +78,44 @@ describe('find* dom-testing-library commands', () => {
...
     })
   })
 
   it('findByText in container', () => {
-    return cy.get('#nested')
-      .then(subject => {
-        cy.findByText(/^Button Text/, {container: subject}).click()
-      })
+    // NOTE: Cypress' `then` doesn't actually return a promise
+    // eslint-disable-next-line jest/valid-expect-in-promise
```
Comment (kentcdodds) [inline cypress/integration/find.spec.js]: We should probably disable the jest plugin for these tests. But we can do that in another PR :)

File: cypress/integration/find.spec.js [hunk 2]
```diff
@@ -30,11 +30,11 @@ describe('find* dom-testing-library commands', () => {
       .click()
       .should('contain', 'Button Clicked')
   })
-  
+
```
Comment (kentcdodds) [inline cypress/integration/find.spec.js]: Sorry about all these. Don't know how those slipped through in the first place

File: cypress/integration/find.spec.js [hunk 3]
```diff
@@ -79,27 +78,44 @@ describe('find* dom-testing-library commands', () => {
...
+    )
+  })
+
+  it('findByText with a previous subject', () => {
+    cy.get('#nested')
+      .findByText('Button Text 1')
+      .should('not.exist')
+    cy.get('#nested')
+      .findByText('Button Text 2')
+      .should('exist')
```
Comment (kentcdodds) [inline cypress/integration/find.spec.js]: Love this

**Comment by kentcdodds** [comment]
@all-contributors please add @tlrobinson for code and tests

**Comment by kentcdodds** [comment]
Thanks so much for your help! I've added you as a collaborator on the project. Please make sure that you review the `other/MAINTAINING.md` and `CONTRIBUTING.md` files (specifically the bit about the commit messages and the git hooks) and familiarize yourself with the code of conduct (we're using [the contributor covenant](http://contributor-covenant.org/)). You might also want to [watch the repo](https://help.github.com/articles/watching-repositories/) to be notified when someone files an issue/PR. Please continue to make PRs as you feel the need (you can make your branches directly on the repo rather than your fork if you want). Thanks! And welcome to the team :)

**Comment by twgraham** [comment]
FYI this was a breaking change - all my tests that assumed findByText used the document broke after upgrading. It might be worth mentioning this in the releases, or bumping the major version if your doing semver.

**Comment by tlrobinson** [comment]
@twgraham out of curiosity, what did your tests that broke look like? This change shouldn’t break queries like `cy.findByText(...)`. It would break some queries like `cy.get(...).findByText(...)`, but IMHO the old behavior was broken.

**Comment by twgraham** [comment]
@tlrobinson kind of like the latter... `cy.get(...).click(...).findBy*(...).click`. E.g. Get this form field and type, then this field and type, then submit, all chained together.

I suppose we were treating a bit like 'get' because we realised early on that the previous subject wasn't yielded, so we had to do something like `cy.get(...).then(el => cy.findByText(..., { container: el }).should(...))` if we wanted to scope it.

I think this PR is definitely a great idea - just didn't want people to get bitten unnecessarily.

**Comment by kentcdodds** [comment]
More discussion on this in #109. Please comment when you're able :)

**Comment by adamtay** [comment]
Great update and definitely prefer this style but as mentioned above, this has introduced a breaking change.

Example usage: `cy.findByText(...).click().findByText('BREAK')`

**Comment by kentcdodds** [comment]
Thanks @adamtay,

We are aware of this. Please see issue #109 where we're discussing what to do about it.
